### ISSUE TITLE
Schema Evolution: Error if numeric type in target has higher precision than in source 

### ISSUE THREAD

**Issue by MarkusRothSBB** [smart-data-lake/smart-data-lake#339]
**Is your feature request related to a problem? Please describe.**
I am ingesting data from a MS SQL database into Snowflake using sdl-snowflake and HistorizeAction. In the source table, there is a column with data type INT (precision 10). Spark converts this to IntegerType (precision 10). I then write this DataFrame to Snowflake, which converts the IntegerType to INTEGER, which is short for NUMERIC(38, 0) in Snowflake. The mapping is done by the [Spark Snowflake Connector](https://docs.snowflake.com/en/user-guide/spark-connector-use.html). This works fine for the first ingest.

During the next ingest, HistorizeAction reads the column with type NUMERIC(38, 0) from Snowflake and converts it to DecimalType(38, 0), again via the Spark Snowflake Connector. Now HistorizeAction invokes SchemaEvolution. SchemaEvolution sees that the old Type (read from Snowflake) is DecimalType(38, 0), and the new type (read from MS SQL) is IntegerType. It concludes that values from DecimalType(38, 0) cannot be converted to IntegerType because the precision in IntegerType is lower and data would be lost. An Exception is thrown in ValueProjector.scala:94.

Therefore, during the second ingest, the feed fails to execute. Even though there is no read danger, because all data comes from MS SQL Server, and so all values are guaranteed to fit into IntegerType.

**Describe the solution you'd like**
I'm not entirely sure. One solution is to cast the values from SQL Server to DecimalType(38, 0) when reading them into a DataFrame using a custom transformer, but that would be necessary for every INT column ingested from MS SQL Server to Snowflake. 
There doesn't seem to be a Snowflake data type that the Spark Snowflake Connector automatically converts to Spark IntegerType.
I'd be happy to get any insight on this issue.

**Comment by MarkusRothSBB** [comment]
This commit fixes the issue: https://github.com/smart-data-lake/smart-data-lake/commit/e84e80225c7bbecd95866c85caf2f8afe1259ca5

Originally, and consistently with the existing conversions, I had used the line:

`      case (d: DecimalType, _: IntegerType) if d.precision <= 16 => (x => x.asInstanceOf[BigDecimal].toDouble)`

But now I changed it to:

`      case (d: DecimalType, _: IntegerType)  => (x => x.asInstanceOf[BigDecimal].toDouble)`

And this particular problem is solved. However, I am not sure if this is generally applicable. 

I guess Schema Evolution wants to protect against the case that the source system reduces the precision of a data type, and then schema evolution reduces the precision of the target column type, thus losing data. But is this a real scenario?

**Comment by zzeekk** [comment]
Hi @MarkusRothSBB, glad you found a workaround!
I have the following improvement to the workaround: it should be .toInt instead of .toDouble at the line end:
`case (d: DecimalType, _: IntegerType) => (x => x.asInstanceOf[BigDecimal].toInt)`

The clean way would be to "cast the values from SQL Server to DecimalType(38, 0)" as suggested by you. This can be done generically, see also https://github.com/smart-data-lake/smart-data-lake/blob/c8d733dbfd64a3ae04da5279d047d527d3d5312b/sdl-core/src/main/scala/io/smartdatalake/util/misc/DataFrameUtil.scala#L82
This will become much more easy & flexible with #344.

On the other side there is potential to extend ValueProject.getSimpleTypeConverter with more conversions, but i'm against the extension suggested by you as it looses precision.
Extension we could make are (but they dont solve your case):
```
case (d: DecimalType, _: ByteType) if d.scale == 0 && d.precision <= 2 => (x => x.asInstanceOf[BigDecimal].toByte)
case (d: DecimalType, _: ShortType) if d.scale == 0 && d.precision <= 4 => (x => x.asInstanceOf[BigDecimal].toShort)
case (d: DecimalType, _: IntegerType) if d.scale == 0 && d.precision <= 9 => (x => x.asInstanceOf[BigDecimal].toInt)
case (d: DecimalType, _: LongType) if d.scale == 0 && d.precision <= 18 => (x => x.asInstanceOf[BigDecimal].toLong)
```

The problem is the current assumption of the SchemaEvolution, that the new dataType must be a superset of the existing dataType. So another way to solve your case would be to extend the SchemaEvolution so it takes the bigger of both dataTypes.

**Comment by MarkusRothSBB** [comment]
Thank you for your insights! The castDfColumnTyp helper method as a first step, and #344 in the long term, will certainly work for me. I still consider them workarounds: It should be possible to import an INT column from SQL server and use the HistorizeAction without having to cast anything. It seems like quite a common use-case.

> another way to solve your case would be to extend the SchemaEvolution so it takes the bigger of both dataTypes.

I think this is the same idea I suggested here:

> For integer/decimal columns where the target has a higher precision than the source, the source column is cast to the target column type.

I think this is the correct solution. No workarounds are needed, and data can't be lost. Without having looked into it very deeply, I do think it would take a considerable refactoring of the Schema Evolution component.

**Comment by pgruetter** [comment]
Had a similar issue: A data field coming from Oracle is Decimal(3,0). At the beginning, it was always null and the Hive field was declared as Decimal(38,10). After a few months, the field in Oracle contains data, Spark correctly defines Decimal(3,0) now so I hit this bug in HistorizeAction: The existing field in Hive is Decimal(38,10), the source field is Decimal(3,0).

**Comment by pgruetter** [comment]
@zzeekk Can you remember if the changes in #397 solve this one completely or if there's something left to do here?

**Comment by zzeekk** [comment]
@pgruetter: no, only a small improvement was implemented to support more mappings of DecimalType to Byte/Short/Int/LongType. What's missing is to ```to extend the SchemaEvolution so it takes the bigger of both dataTypes.```
